### PR TITLE
STM: State Machine Outline

### PR DESCRIPTION
**Pull request by DylanCavers**

### CODE REVIEW

File: lib/state_machine/state_machine.cpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+#include <optional>
+
+namespace hyped::state_machine {
+
+StateMachine::StateMachine()
+{
+}
+
+// TODOLater implement
+std::optional<Message> StateMachine::checkTransition()
```
Comment (mifrandir) [inline lib/state_machine/state_machine.cpp:13]: do we have a better name?

File: lib/state_machine/state_machine.cpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  /*
+  check if a transition is needed with current data
+  then run transition to move to next state
+  return transition message or no transition
+  */
+  return std::nullopt;
+}
+
+// Transition to next state
+void StateMachine::transition(const Message message)
```
Comment (mifrandir) [inline lib/state_machine/state_machine.cpp:24]: `handleMessage`?

File: lib/state_machine/state_machine.hpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  StateMachine() { current_state = State::kIdle; };
+
+  void transition(const Message message);
+
+  std::optional<Message> checkTransition();
+
+  void reset();
+
+  // TODO change transitions to actual transitions
+  std::array<Transition, kNumTransitions> transitions{
```
Comment (mifrandir) [inline lib/state_machine/state_machine.hpp:23]: I don't agree with the data structure here.

Think about "What do I want to do?". I claim there is exactly one way in which you are going to look things up. The current lookup time is `O(n * m)` where `n` is the number of states and `m` the number of messages; this should be constant. Talk to my IADS tutees, they should know how to make this better.

File: lib/state_machine/types.hpp [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+#include "message.hpp"
+#include "state.hpp"
+
+namespace hyped::state_machine {
+
+struct Transition {
+  State from;
+  State to;
+  Message message;
```
Comment (mifrandir) [inline lib/state_machine/types.hpp:11]: source, target, message?

something like `lookup(transition.to, transition.message)` doesn't read well, I think.

I am also not convinced we need this struct at all but I'll see.

**Review by mifrandir — CHANGES_REQUESTED** [review]
Remove the `.DS_Store`

File: lib/state_machine/state_machine.hpp (hunk 3 above)
Comment (DylanCavers) [reply lib/state_machine/state_machine.hpp:23]: Would an unordered map be more appropriate for this?

Comment (mifrandir) [reply lib/state_machine/state_machine.hpp:23]: certainly

you probably want to use a key like `State * Message` but put that in a struct like `Event` and then figure out how to hash it.

Comment (kshxtij) [reply lib/state_machine/state_machine.hpp]: For the hashing, i think we can just copy over Boost's hash_combine() and put it in utils::hash.

Comment (mifrandir) [reply lib/state_machine/state_machine.hpp]: I mean we could also just actually depend on boost, like approximately any other C++ project :P

Comment (mifrandir) [reply lib/state_machine/state_machine.hpp]: But there's another thing: you either need to make a custom hasher and carry it around or invoke some black magic which I unfortunately forgot (it involved defining an empty template inside the `std` namespace).

Ofc you might say "but why not use `std::pair<S,T>` and the answer is because the actual code (outside the types) will be ugly if we do this.

File: lib/state_machine/state_machine.cpp [hunk 5]
```diff
@@ -0,0 +1,34 @@
+#include "state_machine.hpp"
+#include "types.hpp"
+
+#include <optional>
```
Comment (maxguy2001) [inline lib/state_machine/state_machine.cpp:4]: Already included in header

File: lib/state_machine/types.hpp [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  bool operator==(const TransitionKey &key) const
+  {
+    return key.source == source && key.message == message;
+  }
+};
+
+struct TransitionHasher {
+  std::size_t operator()(const TransitionKey &key) const
+  {
+    using std::size_t;
```
Comment (maxguy2001) [inline lib/state_machine/types.hpp:21]: ?

File: lib/state_machine/types.hpp [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+    return ((int)key.source);
```
Comment (maxguy2001) [inline lib/state_machine/types.hpp:23]: `static_cast<int>(key.source)`?

**Review by maxguy2001 — CHANGES_REQUESTED** [review]

File: lib/state_machine/state_machine.cpp [hunk 8]
```diff
@@ -0,0 +1,28 @@
+#include "state_machine.hpp"
+#include "types.hpp"
```
Comment (TomLonergan03) [inline lib/state_machine/state_machine.cpp:2]: already included in header

File: lib/state_machine/state_machine.hpp [hunk 9]
```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include "consts.hpp"
+#include "state.hpp"
+#include "types.hpp"
+
+#include <array>
```
Comment (TomLonergan03) [inline lib/state_machine/state_machine.hpp:7]: unused as far as i can see

File: lib/state_machine/state_machine.hpp [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+       {{State::kMitigate, Message::mPostEmergency}, State::kPECruising},
+       {{State::kMitigate, Message::mPostEmergency}, State::kPEBraking},
+       {{State::kMitigate, Message::mPostEmergency}, State::kMotorBraking},
+       {{State::kMitigate, Message::mPostEmergency}, State::kFrictionBraking},
+       {{State::kMitigate, Message::mPostEmergency}, State::kFailureBraking},
+       {{State::kPEAccelerating, Message::mResume}, State::kAccelerating},
+       {{State::kPECruising, Message::mResume}, State::kCruising},
+       {{State::kPEBraking, Message::mResume}, State::kNominalBraking}};
+
+  State current_state;
```
Comment (TomLonergan03) [inline lib/state_machine/state_machine.hpp:70]: `current_state_`, also should probably be private, with a getter function if needed

**Review by TomLonergan03 — CHANGES_REQUESTED** [review]

File: lib/state_machine/state_machine.hpp (hunk 3 above)
Comment (mifrandir) [reply lib/state_machine/state_machine.hpp]: that's the magic i was talking about

File: lib/utils/hash.hpp [hunk 11]
```diff
@@ -0,0 +1,12 @@
+namespace hyped::utils {
+
+inline void hash_combine(std::size_t& seed) { }
+
+template <class T>
+inline void hash_combine(std::size_t & seed, const T& v)
+{
+  std::hash<T> hasher;
+  seed ^= hasher(v) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
+};
```
Comment (mifrandir) [inline lib/utils/hash.hpp]: I don't like this snippet because it looks so arbitrary; can we not just depend on boost? Just require it to be present on the host like we do with eigen and we should be good. @kshxtij

File: lib/state_machine/types.hpp [hunk 12]
```diff
@@ -18,9 +20,10 @@ struct TransitionKey {
 struct TransitionHasher {
```
Comment (mifrandir) [inline lib/state_machine/types.hpp:22]: We would rather do the thing where we add a hash function in `std::` than use a hasher like this.

File: lib/state_machine/types.hpp [hunk 13]
```diff
@@ -18,9 +20,10 @@ struct TransitionKey {
...
   std::size_t operator()(const TransitionKey &key) const
```
Comment (mifrandir) [inline lib/state_machine/types.hpp:23]: `TransitionKey` is a bad name; what does it contain? If there's no more meaningful name call it `SourceAndMessage`

**Review by mifrandir — CHANGES_REQUESTED** [review]
remove `lib/.DS_Store`

File: lib/state_machine/types.hpp [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+#include "message.hpp"
+#include "state.hpp"
+
+#include <cstdint>
+
+#include "utils/hash.hpp"
+
+namespace hyped::state_machine {
+
+struct TransitionKey {
```
Comment (mifrandir) [inline lib/state_machine/types.hpp:12]: This could just live inside `state_machine::StateMachine`? It's not going to be used anywhere else and it's just for the internals.

I'm not sure whether that's possible, but I believe you can make it private. Might be an issue with the `std::` template, though...

File: lib/state_machine/types.hpp [hunk 15]
```diff
@@ -0,0 +1,33 @@
+#pragma once
+
+#include "message.cpp"
```
Comment (maxguy2001) [inline lib/state_machine/types.hpp:3]: Include headers, not cpp files

**Review by maxguy2001 — CHANGES_REQUESTED** [review]

File: test/state_machine/stm_tests.cpp [hunk 16]
```diff
@@ -0,0 +1,5 @@
+#include <gtest/gtest.h>
+
+namespace hyped::test {
+
```
Comment (mifrandir) [inline test/state_machine/stm_tests.cpp:4]: todo?

File: lib/state_machine/state_machine.hpp [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+ public:
+  StateMachine() { current_state = State::kIdle; };
+
+  void handleMessage(const Message message);
+
+  std::optional<Message> handleData();
+
+  void reset();
+
+  const std::unordered_map<SourceAndMessage, State> transitions
```
Comment (maxguy2001) [inline lib/state_machine/state_machine.hpp:22]: Should this definitely be public?

**Review by maxguy2001 — CHANGES_REQUESTED** [review]

File: lib/state_machine/types.hpp [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+  bool operator==(const SourceAndMessage &key) const
+  {
+    return key.source == source && key.message == message;
+  }
+};
+}  // namespace hyped::state_machine
+
+namespace std {
+template<>
+struct hash<hyped::state_machine::SourceAndMessage> {
```
Comment (maxguy2001) [inline lib/state_machine/types.hpp:24]: redundant `hyped`

Comment (maxguy2001) [reply lib/state_machine/types.hpp:24]: And elsewhere in the struct.

**Review by maxguy2001 — CHANGES_REQUESTED** [review]

File: lib/state_machine/state_machine.hpp (hunk 17 above)
Comment (mifrandir) [reply lib/state_machine/state_machine.hpp:22]: no!

File: test/state_machine/stm_tests.cpp [hunk 19]
```diff
@@ -0,0 +1,5 @@
+#include <gtest/gtest.h>
+
+namespace hyped::test {
+// TODO transition testing
```
[the remark this answers is not in the document]
Comment (mifrandir) [reply test/state_machine/stm_tests.cpp:4]: nope, `TODO`; let them fix it in this PR

File: lib/state_machine/state_machine.cpp [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+  return std::nullopt;
+}
+
+// Transition to next state
+void StateMachine::handleMessage(const Message message)
+{
+  current_state = transitions.at({current_state, message});
+}
+
+void StateMachine::reset()
```
[the remark this answers is not in the document]
Comment (mifrandir) [reply lib/state_machine/state_machine.cpp:22]: I agree. In fact I don't like this function at all. You should get rid of it in order to avoid bugs. The user can construct a new state machine if that's what's required.

File: lib/state_machine/state_machine.hpp (hunk 17 above)
Comment (mifrandir) [reply lib/state_machine/state_machine.hpp:22]: also naming

File: lib/state_machine/state_machine.hpp [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+
+namespace hyped::state_machine {
+
+class StateMachine {
+ public:
+  StateMachine() { current_state = State::kIdle; };
+
+  void handleMessage(const Message message);
+
+  std::optional<Message> handleData();
```
Comment (mifrandir) [inline lib/state_machine/state_machine.hpp:18]: what is this function supposed to do? it says handle but there isn't any input?

File: lib/state_machine/types.hpp (hunk 18 above)
Comment (maxguy2001) [reply lib/state_machine/types.hpp:24]: Then we'll have to change the job to check for redundant hyped since this currently causes it to fail.

Comment (mifrandir) [reply lib/state_machine/types.hpp:24]: @kshxtij :joy:

File: lib/state_machine/state_machine.hpp (hunk 21 above)
Comment (DylanCavers) [reply lib/state_machine/state_machine.hpp:19]: unsure on the input at the moment, aim is for this function to handle the messages received from other the other nodes and transition appropriately, will try figure something out for this on Saturday.

File: lib/state_machine/state_machine.hpp [hunk 22]
```diff
@@ -17,8 +19,18 @@ class StateMachine {
...
+
  private:
-  const std::unordered_map<SourceAndMessage, State> transitions
+  const std::unordered_map<std::string, Message> string_to_message
+    = {{"mForward", Message::mForward}, {"mFailure", Message::mFailure}};
+
+  const std::unordered_map<Message, std::string> message_to_string
+    = {{Message::mForward, "mForward"}, {Message::mFailure, "mFailure"}};
+
+  const boost::unordered_map<SourceAndMessage, State> transitionMap
```
Comment (mifrandir) [inline lib/state_machine/state_machine.hpp:33]: naming style

File: lib/state_machine/state_machine.hpp [hunk 23]
```diff
@@ -17,8 +19,18 @@ class StateMachine {
 
   std::optional<Message> handleData();
 
+  Message stringToMessage(const std::string message_name);
+
+  std::string messageToString(const Message message);
```
Comment (mifrandir) [inline lib/state_machine/state_machine.hpp:24]: references

**Review by mifrandir — CHANGES_REQUESTED** [review]

File: lib/state_machine/state_machine.hpp [hunk 24]
```diff
@@ -19,18 +19,18 @@ class StateMachine {
...
-  std::string messageToString(const Message message);
+  std::string messageToString(const Message &message);
 
  private:
-  const std::unordered_map<std::string, Message> string_to_message
+  const std::unordered_map<std::string, Message> StringMap
     = {{"mForward", Message::mForward}, {"mFailure", Message::mFailure}};
 
-  const std::unordered_map<Message, std::string> message_to_string
+  const std::unordered_map<Message, std::string> MessageMap
```
Comment (mifrandir) [inline lib/state_machine/state_machine.hpp:30]: What happened here?

This should be `message_to_string_` or `kMessageToString`

similar for the others

File: lib/state_machine/state_machine.cpp [hunk 25]
```diff
@@ -0,0 +1,34 @@
+#include "state_machine.hpp"
+
+namespace hyped::state_machine {
+
+// TODOLater implement
+std::optional<Message> StateMachine::handleData()
```
Comment (maxguy2001) [inline lib/state_machine/state_machine.cpp:6]: Should this not be `checkTransiton` or similar?

Comment (maxguy2001) [reply lib/state_machine/state_machine.cpp:6]: The name just doesn't seem to describe the functionality of this?

**Review by maxguy2001 — CHANGES_REQUESTED** [review]

File: lib/state_machine/state_machine.cpp (hunk 25 above)
Comment (DylanCavers) [reply lib/state_machine/state_machine.cpp:6]: https://github.com/Hyp-ed/hyped-2023/pull/69#discussion_r1083266641

File: lib/state_machine/message.hpp [hunk 26]
```diff
@@ -0,0 +1,15 @@
+#pragma once
+
+namespace hyped::state_machine {
+
+// State class containing all the states
+enum class Message {
+  mForward,
```
Comment (TomLonergan03) [inline lib/state_machine/message.hpp:7]: what is `mForward`? is there a better name for this or is this placeholder messages?

File: lib/state_machine/state_machine.cpp (hunk 25 above)
Comment (TomLonergan03) [reply lib/state_machine/state_machine.cpp:6]: I prefer `checkTransition` to `handleData` tbh

File: lib/state_machine/state_machine.cpp [hunk 27]
```diff
@@ -0,0 +1,34 @@
...
+// Transition to next state
+void StateMachine::handleMessage(const Message message)
+{
+  current_state = transition_to_state_.at({current_state, message});
+}
+
+Message StateMachine::stringToMessage(const std::string &message_name)
+{
+  Message message = string_to_message_.at(message_name);
+  return message;
```
Comment (TomLonergan03) [inline lib/state_machine/state_machine.cpp:25]: `return string_to_message_.at(message_name);`

File: lib/state_machine/state_machine.cpp [hunk 28]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+std::string StateMachine::messageToString(const Message &message)
+{
+  std::string message_name = message_to_string_.at(message);
+  return message_name;
```
Comment (TomLonergan03) [inline lib/state_machine/state_machine.cpp:31]: `return message_to_string_.at(message);`

File: lib/state_machine/state_machine.hpp (hunk 21 above)
Comment (TomLonergan03) [reply lib/state_machine/state_machine.hpp:20]: if no progress has happened with this, add a TODOLater

File: lib/state_machine/message.hpp (hunk 26 above)
Comment (TomLonergan03) [reply lib/state_machine/message.hpp:7]: looking at the rest of the PR this would be better as something like `kNextNominalState`

**Review by TomLonergan03 — CHANGES_REQUESTED** [review]

File: lib/state_machine/state_machine.cpp [hunk 29]
```diff
@@ -0,0 +1,32 @@
...
+  /*
+  check if a transition is needed with current data from ros node
+  then run transition to move to next state
+  return transition message or no transition
+  */
+  return std::nullopt;
+}
+
+// Transition to next state
+void StateMachine::handleMessage(const Message message)
```
Comment (TomLonergan03) [inline lib/state_machine/state_machine.cpp:17]: should this take a reference?

File: lib/state_machine/state_machine.hpp [hunk 30]
```diff
@@ -0,0 +1,61 @@
...
+       {{State::kNominalBraking, Message::kFailureBrake}, State::kFailureBraking},
+       {{State::kMotorBraking, Message::kFailureBrake}, State::kFailureBraking},
+       {{State::kMotorBraking, Message::kFrictionBrake}, State::kFrictionBraking},
+       {{State::kMotorBraking, Message::kFailure}, State::kFailureStopped},
+       {{State::kFrictionBraking, Message::kFailureBrake}, State::kFailureBraking},
+       {{State::kFrictionBraking, Message::kFailure}, State::kFailureStopped},
+       {{State::kFailureBraking, Message::kFailure}, State::kFailureStopped},
+       {{State::kFailureStopped, Message::kFailure}, State::kOff}};
+
+  State current_state;
```
Comment (TomLonergan03) [inline lib/state_machine/state_machine.hpp:59]: `current_state_`

File: lib/state_machine/state_machine.hpp [hunk 31]
```diff
@@ -0,0 +1,61 @@
...
+#include <optional>
+#include <unordered_map>
+
+#include <boost/unordered_map.hpp>
+
+namespace hyped::state_machine {
+
+class StateMachine {
+ public:
+  StateMachine() { current_state = State::kIdle; };
```
Comment (TomLonergan03) [inline lib/state_machine/state_machine.hpp:16]: constructor should not be in the header, also initialise `current_state_` in the initialiser list

**Review by TomLonergan03 — CHANGES_REQUESTED** [review]

File: lib/state_machine/state_machine.cpp (hunk 25 above)
Comment (kshxtij) [reply lib/state_machine/state_machine.cpp]: @mifrandir what was the problem with `checkTransition`?

**Review by TomLonergan03 — APPROVED** [review]
Looks good

File: lib/state_machine/message.hpp [hunk 32]
```diff
@@ -0,0 +1,15 @@
+#pragma once
+
+namespace hyped::state_machine {
+
+// State class containing all the states
```
Comment (SnickeyX) [inline lib/state_machine/message.hpp:5]: State class?

File: lib/state_machine/state_machine.hpp [hunk 33]
```diff
@@ -0,0 +1,59 @@
...
+  StateMachine();
+
+  // TODOLater handle ros messages and transition accordingly
+  std::optional<Message> checkTransition();
+
+  void handleMessage(const Message &message);
+
+  Message stringToMessage(const std::string &message_name);
+
+  std::string messageToString(const Message &message);
```
Comment (SnickeyX) [inline lib/state_machine/state_machine.hpp:24]: if there aren't any comments, you can leave out the whitespace

File: lib/state_machine/state_machine.hpp [hunk 34]
```diff
@@ -0,0 +1,59 @@
...
+       {{State::kMotorBraking, Message::kPreFrictionBrake}, State::kPreFrictionBraking},
+       {{State::kPreFrictionBraking, Message::kFrictionBrake}, State::kFrictionBraking},
+       {{State::kFrictionBraking, Message::kFailureBrake}, State::kFailureBraking},
+       {{State::kFrictionBraking, Message::kFailure}, State::kFailureStopped},
+       {{State::kFailureBraking, Message::kFailure}, State::kFailureStopped},
+       {{State::kFailureStopped, Message::kFailure}, State::kOff}};
+
+  State current_state_;
+};
+}  // namespace hyped::state_machine
```
Comment (SnickeyX) [inline lib/state_machine/state_machine.hpp:59]: space before last bracket (pedantic ik)

File: lib/state_machine/types.hpp [hunk 35]
```diff
@@ -0,0 +1,28 @@
...
+};
+
+std::size_t hash_value(SourceAndMessage const &key)
+{
+  std::size_t seed = 0;
+  boost::hash_combine(seed, key.message);
+  boost::hash_combine(seed, key.source);
+  return seed;
+}
+}  // namespace hyped::state_machine
```
Comment (SnickeyX) [inline lib/state_machine/types.hpp:28]: newline before last bracket

File: test/state_machine/stm_tests.cpp (hunk 19 above)
Comment (SnickeyX) [inline test/state_machine/stm_tests.cpp:4]: TODOLater since I'm assuming this is going to be a different PR?

**Review by SnickeyX — CHANGES_REQUESTED** [review]
only formatting changes

File: lib/state_machine/state.hpp [hunk 36]
```diff
@@ -0,0 +1,21 @@
+#pragma once
+
+namespace hyped::state_machine {
+
+// State class containing all the states
```
Comment (SnickeyX) [inline lib/state_machine/state.hpp:5]: this comment feels a bit redundant given the context of state machine

**Review by SnickeyX — CHANGES_REQUESTED** [review]
a lil more, will approve right after

**Review by SnickeyX — APPROVED** [review]
wheyyy

File: lib/state_machine/message.hpp [hunk 37]
```diff
@@ -0,0 +1,15 @@
+#pragma once
+
+namespace hyped::state_machine {
+
+// Message class containing messages that prompt transitions
+enum class Message {
+  kNextNominalState,
```
Comment (mifrandir) [inline lib/state_machine/message.hpp:7]: How are you going to make sure that receiving two of these in a row is not going to advance the state machine too far? Like, how do you avoid one of the controllers sending this message twice or, more realistically, two individual controllers sending this message and the receiver transitioning twice (instead of the intended once).

Comment (DylanCavers) [reply lib/state_machine/message.hpp:7]: Was initially decided on to allow different transition orders to end up in the same state, but I guess this doesn't work as well for the NextNominalState transitions as if two messages are sent from different controller to transition from the same state, they would transition too far and essentially skip a state. This would be needed for failure states to ensure that no matter the order of received messages the failure transition is always made. So might be better to add separate messages for all the non-failure transitions, but keep the general transitions for failure states?

Comment (mifrandir) [reply lib/state_machine/message.hpp:7]: That may be a good idea, yes.

**Review by kshxtij — APPROVED** [review]
